### PR TITLE
Introduce the `pkg/bitcoin` package

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/keep-core/issues/3349

Here we introduce the new `pkg/bitcoin` package that defines some basic Bitcoin types and interfaces required to work with the Bitcoin chain. This package is meant to reflect the part of the Bitcoin protocol to an extent required by the client applications. The `pkg/bitcoin` package is supposed to hold components specific to the Bitcoin domain and must remain free of application-specific items. Third-party helper libraries (like `btcd`) are allowed for use within this package though no external type should leak outside. Preserving those rules will allow us to achieve a clean package structure with proper dependencies. The exact types introduced by this pull request are:
- `bitcoin.ByteOrder`: An enum that captures all the caveats related to the byte order used across the Bitcoin ecosystem
- `bitcoin.CompactSizeUint`: A documentation type that holds the details about the CompactSize Unsigned Integer
- `bitcoin.Hash`: A type that represents a 32-byte Bitcoin hash widely used as identifier in the Bitcoin protocol
- `bitcoin.Transaction` & related: A type that represents a Bitcoin transaction and related types to reflect the internals
- `bitcoin.BlockHeader`: A representation of the Bitcoin block header that will be used for SPV purposes
- `bitcoin.Chain`: A minimal interface required to interact with the Bitcoin chain
- `electrum.Client`: An example stub implementation of the `bitcoin.Chain` supposed to establish a package structure for Bitcoin interfaces implementations

### CODE REVIEW

File: pkg/bitcoin/bitcoin.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	// protocol. This is the primary byte order that is suitable for the
+	// use cases related with the protocol logic and cryptography. Byte arrays
+	// using this byte order should be converted to numbers using according to
+	// the little-endian sequence.
+	InternalByteOrder = iota
+
+	// ReversedByteOrder represents the "human" byte order. This is the
+	// byte order that is typically used by the third party services like
+	// block explorers or Bitcoin chain clients. Byte arrays using this byte
+	// order should be converted to numbers using  according to the big-endian
```
Comment (tomaszslabon) [inline pkg/bitcoin/bitcoin.go:31]:
```suggestion
	// order should be converted to numbers according to the big-endian
```

File: pkg/bitcoin/chain.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	// transaction with the given transaction hash. If the transaction with the
+	// given hash was not found on the chain, this function returns an error.
+	GetTransactionConfirmations(transactionHash Hash) (uint, error)
+
+	// BroadcastTransaction broadcasts the given transaction over the
+	// network of the Bitcoin chain nodes. If the broadcast action could not be
+	// done, this function returns an error. This function does not give any
+	// guarantees regarding transaction mining. The transaction may be mined or
+	// rejected eventually.
+	BroadcastTransaction(transactionHash Hash) error
```
Comment (tomaszslabon) [inline pkg/bitcoin/chain.go:21]: `BroadcastTransaction` should probably accept raw data of transaction as the argument (`[]byte` or some custom type allowing to pass any number of bytes representing the transaction).

File: pkg/bitcoin/transaction.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	//               as an CompactSizeUint.
+	//       3. Serialize t.Outputs just like t.Inputs.
+	//       4. Serialize t.Locktime to an InternalByteOrder byte array.
+	return nil
+}
+
+// Hash calculates the transaction's hash as the double SHA-256 of the
+// traditional serialization format: [version][inputs][outputs][locktime].
+// The outcome is equivalent to the txid field defined in the Bitcoin
+// specification. Do not confuse it with the wtxid field defined by BIP 141.
```
Comment (tomaszslabon) [inline pkg/bitcoin/transaction.go:39]:
```suggestion
// specification. Do not confuse it with the txid field defined by BIP 141.
```

File: pkg/bitcoin/hash.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+type Hash [HashByteLength]byte
+
+// NewHashFromString creates a new Hash instance using the given string.
+// The string is interpreted according to the given ByteOrder. That is, the
+// string is taken as is if the ByteOrder is InternalByteOrder and reversed if
+// the ByteOrder is ReversedByteOrder. The string's length must be equal
+// to 2*HashByteLength.
+func NewHashFromString(hash string, byteOrder ByteOrder) (Hash, error) {
+	if len(hash) != 2*HashByteLength {
+		return [HashByteLength]byte{}, fmt.Errorf("wrong hash string size")
```
Comment (tomaszslabon) [inline pkg/bitcoin/hash.go:22]:
```suggestion
		return Hash{}, fmt.Errorf("wrong hash string size")
```

File: pkg/bitcoin/hash.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	hashBytes, err := hex.DecodeString(hash)
+	if err != nil {
+		return [HashByteLength]byte{}, fmt.Errorf(
```
Comment (tomaszslabon) [inline pkg/bitcoin/hash.go:27]:
```suggestion
		return Hash{}, fmt.Errorf(
```

File: pkg/bitcoin/hash.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// NewHash creates a new Hash instance using the given byte slice.
+// The byte slice is interpreted according to the given ByteOrder. That is, the
+// byte slice is taken as is if the ByteOrder is InternalByteOrder and reversed
+// if the ByteOrder is ReversedByteOrder. The byte slice's length must be equal
+// to HashByteLength.
+func NewHash(hash []byte, byteOrder ByteOrder) (Hash, error) {
+	if len(hash) != HashByteLength {
+		return [HashByteLength]byte{}, fmt.Errorf("wrong hash size")
```
Comment (tomaszslabon) [inline pkg/bitcoin/hash.go:43]:
```suggestion
		return Hash{}, fmt.Errorf("wrong hash size")
```

File: pkg/bitcoin/transaction.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+// For reference, see:
+// https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-id
+func (t *Transaction) Hash() Hash {
+	// TODO: Implementation of the Hash function that consists of the following:
+	//       1. Call t.Serialize() to get the serialized transaction.
+	//       2. Compute the double SHA-256 over the serialized transaction.
+	//       3. Construct the Hash instance appropriately.
+	return [HashByteLength]byte{}
```
Comment (tomaszslabon) [inline pkg/bitcoin/transaction.go:47]:
```suggestion
	return Hash{}
```

File: pkg/bitcoin/chain.go (hunk 2 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/chain.go]: Yep, my bad! Good catch, thanks! See 235dddf8

File: pkg/bitcoin/bitcoin.go (hunk 1 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/bitcoin.go]: See 38d7930c

File: pkg/bitcoin/hash.go (hunk 4 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/hash.go]: See bf124495

File: pkg/bitcoin/hash.go (hunk 5 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/hash.go]: See bf124495

File: pkg/bitcoin/block.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+// Hash calculates the block header's hash as the double SHA-256 of the
+// block header serialization format:
+// [Version][PreviousBlockHeaderHash][MerkleRootHash][Time][Bits][Nonce].
+func (bh *BlockHeader) Hash() Hash {
+	// TODO: Implementation of the Hash function that consists of the following:
+	//       1. Call bh.Serialize() to get the serialized block hash.
+	//       2. Compute the double SHA-256 over the serialized  block hash.
+	//       3. Construct the Hash instance appropriately.
+	return [HashByteLength]byte{}
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/bitcoin/block.go]: See bf124495

File: pkg/bitcoin/hash.go (hunk 6 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/hash.go]: See bf124495

File: pkg/bitcoin/transaction.go (hunk 7 above)
Comment (lukasz-zimnoch) [reply pkg/bitcoin/transaction.go]: See bf124495

**Review by tomaszslabon — DISMISSED** [review]

**Comment by tomaszslabon** [comment]
Looks good to me now.
I leave it to @pdyraga and @nkuba.

File: pkg/bitcoin/bitcoin.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	// use cases related with the protocol logic and cryptography. Byte arrays
+	// using this byte order should be converted to numbers according to
+	// the little-endian sequence.
+	InternalByteOrder = iota
+
+	// ReversedByteOrder represents the "human" byte order. This is the
+	// byte order that is typically used by the third party services like
+	// block explorers or Bitcoin chain clients. Byte arrays using this byte
+	// order should be converted to numbers according to the big-endian
+	// sequence.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/bitcoin/bitcoin.go]: See 2af23fe4

File: pkg/bitcoin/block.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	PreviousBlockHeaderHash Hash
+	// MerkleRootHash is a hash derived from the hashes of all transactions
+	// included in this block.
+	MerkleRootHash Hash
+	// Time is a Unix epoch time when the block was created.
+	Time uint32
+	// Bits determines the target threshold this block's header hash must be
+	// less than or equal to.
+	Bits uint32
+	// Nonce is the nonce used to generate the block.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/bitcoin/block.go]: See 2af23fe4

File: pkg/bitcoin/block.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+type BlockHeader struct {
+	// Version is the block version number that indicates which set of block
+	// validation rules to follow.
+	Version int32
+	// PreviousBlockHeaderHash is the hash of the previous block's header.
+	PreviousBlockHeaderHash Hash
+	// MerkleRootHash is a hash derived from the hashes of all transactions
+	// included in this block.
+	MerkleRootHash Hash
+	// Time is a Unix epoch time when the block was created.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/bitcoin/block.go]: See 2af23fe4

File: pkg/bitcoin/bitcoin.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+// byte orders are used depending on the purpose.
+type ByteOrder int
+
+const (
+	// InternalByteOrder represents the internal byte order used by the Bitcoin
+	// protocol. This is the primary byte order that is suitable for the
+	// use cases related with the protocol logic and cryptography. Byte arrays
+	// using this byte order should be converted to numbers according to
+	// the little-endian sequence.
+	InternalByteOrder = iota
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/bitcoin/bitcoin.go]: Indeed! Good catch! See d31a350b

**Review by nkuba — APPROVED** [review]

**Review by tomaszslabon — APPROVED** [review]

File: pkg/bitcoin/chain.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	// BroadcastTransaction broadcasts the given transaction over the
+	// network of the Bitcoin chain nodes. If the broadcast action could not be
+	// done, this function returns an error. This function does not give any
+	// guarantees regarding transaction mining. The transaction may be mined or
+	// rejected eventually.
+	BroadcastTransaction(transaction *Transaction) error
+
+	// GetCurrentBlockNumber gets the number of the current block. If the
+	// current block was not determined, this function returns an error.
+	GetCurrentBlockNumber() (uint, error)
```
Comment (nkuba) [inline pkg/bitcoin/chain.go:25]: @lukasz-zimnoch should it return the latest mined (confirmed) block or the block that is currently being mined?

Comment (nkuba) [reply pkg/bitcoin/chain.go:25]: To be more precise maybe we could name this function `GetLatestBlockHeight`?

**Pull request merged**